core/primitives/src/stateless_validation/state_witness.rs

```diff
@@ -146,25 +173,50 @@ pub struct ChunkStateWitness {
     /// After these are applied as well, we should arrive at the pre-state-root
     /// of the chunk that this witness is for.
     pub implicit_transitions: Vec<ChunkStateTransition>,
-    #[deprecated(
-        note = "Was used for protocol versions without relaxed chunk validation which is not supported anymore."
-    )]
-    pub _deprecated_new_transactions: Vec<SignedTransaction>,
-    /// The following two field is deprecated and should not be use anymore
-    #[deprecated(
-        note = "Was used for protocol versions without relaxed chunk validation which is not supported anymore."
-    )]
-    pub _deprecated_new_transactions_validation_state: PartialState,
-    // TODO(stateless_validation): Deprecate this field in the next version of the state witness.
-    signature_differentiator: SignatureDifferentiator,
+}
+
+// Note: After deprecation of VersionedStateWitness protocol feature, we can remove this
+impl BorshSerialize for ChunkStateWitness {
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        match self {
+            ChunkStateWitness::V1(witness) => BorshSerialize::serialize(&witness, writer),
+            ChunkStateWitness::V2(witness) => {
+                BorshSerialize::serialize(&1_u8, writer)?;
+                BorshSerialize::serialize(&witness, writer)
+            }
+        }
+    }
+}
+
+// Note: After deprecation of VersionedStateWitness protocol feature, we can remove this
+impl BorshDeserialize for ChunkStateWitness {
+    /// Deserialize based on the first and second bytes of the stream. For V0, we do backward compatible deserialization by deserializing
+    /// the entire stream into V0. For V1, we consume the first byte and then deserialize the rest.
+    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
+        // This is a ridiculous hackery: because the first field in `ChunkStateWitnessV1` is an `AccountId`
+        // and an account id is at most 64 bytes, for all valid `ChunkStateWitnessV1` the second byte must be 0
+        // because of the little endian encoding of the length of the account id.
```

Comment:
I see you've taken inspiration from the receipts serialization ;) This hack is now becoming the norm 🤡 

Can you double check that this invariant holds? This account id is deprecated so I wouldn't be surprised if it was set to some unusual magic value.

Reply:
I like @Longarithm 's idea "As it has epoch_id, we know protocol version of state witness, so we know which version we should decode. Seems cleaner than looking into the raw bytes."

Let me try to see if I can remove the custom serialize/deserialize altogether based on protocol_version for epoch_id. Still not sure how hard it might be.

Reply:
Removed custom serialization/deserialization!